dftimewolf/lib/collectors/grr_hosts.py

```diff
@@ -980,64 +981,85 @@ def _DownloadResults(self,
       results.append(data_frame)
     return results
 
+  def _ProcessQuery(
+      self,
+      hostname: str,
+      osquery_container: containers.OsqueryQuery
+  ) -> None:
+    """Processes an osquery flow on a GRR client.
+
+    Args:
+      hostname (str): the GRR client hostname.
+      osquery_container (containers.OsqueryQuery): the osquery.
+    """
+    client = self._GetClientBySelector(hostname)
+
+    hunt_args = osquery_flows.OsqueryFlowArgs(
+        query=osquery_container.query,
+        timeout_millis=self.timeout_millis,
+        ignore_stderr_errors=self.ignore_stderr_errors)
+
+    try:
+      flow_id = self._LaunchFlow(client, 'OsqueryFlow', hunt_args)
+      self._AwaitFlow(client, flow_id)
+    except DFTimewolfError as error:
+      self.ModuleError(
+          f'Error raised while launching/awaiting flow: {error.message}')
+      return
+
+    name = osquery_container.name
+    description = osquery_container.description
+    query = osquery_container.query
+    flow_identifier = flow_id
+    client_identifier = client.client_id
+
+    results = self._DownloadResults(client, flow_id)
+    if not results:
+      results_container = containers.OsqueryResult(
+          name=name,
+          description=description,
+          query=query,
+          hostname=hostname,
+          data_frame=pd.DataFrame(),
+          flow_identifier=flow_identifier,
+          client_identifier=client_identifier)
+      self.state.StoreContainer(results_container)
+      return
+
+    for data_frame in results:
+      self.logger.info(
+          f'{str(flow_id)} ({hostname}): {len(data_frame)} rows collected')
+
+      dataframe_container = containers.OsqueryResult(
+          name=name,
+          description=description,
+          query=query,
+          hostname=hostname,
+          data_frame=data_frame,
+          flow_identifier=flow_identifier,
+          client_identifier=client_identifier)
+
+      self.state.StoreContainer(dataframe_container)
+
   def Process(self, container: containers.Host) -> None:
     """Collect osquery results from a host with GRR.
 
     Raises:
       DFTimewolfError: if no artifacts specified nor resolved by platform.
     """
-    for client in self._FindClients([container.hostname]):
-      osquery_containers = self.state.GetContainers(containers.OsqueryQuery)
+    osquery_containers = self.state.GetContainers(containers.OsqueryQuery)
 
+    host_osquery_futures = []
+    with ThreadPoolExecutor() as executor:
```

Comment:
What's the default max threads here? Should we limit it to e.g. 5 if we're concerned about not running too many flows concurrently on machines?

Reply:
So running this on my machine (8 cores), means that we'll be scheduling 40 OSQueryFlows on the target machines. Maybe that's a tad too much to leave a default value?

Reply:
FWIW, as of 3.8 max_workers is now ```min(32, os.cpu_count() + 4)```.

I've set this arbitrarily to 4 for now but I also believe GRR will throttle the flows anyway so it doesn't overflow the client.